src/app/codegen-data-model-provider/CodegenDataModelProvider.cpp

```diff
@@ -299,6 +314,87 @@ unsigned FindNextDeviceTypeIndex(Span<const EmberAfDeviceType> types, const Data
 
 const ConcreteCommandPath kInvalidCommandPath(kInvalidEndpointId, kInvalidClusterId, kInvalidCommandId);
 
+std::optional<DataModel::EndpointInfo> GetEndpointInfoAtIndex(uint16_t endpointIndex)
+{
+    VerifyOrReturnValue(emberAfEndpointIndexIsEnabled(endpointIndex), std::nullopt);
+    EndpointId parent = emberAfParentEndpointFromIndex(endpointIndex);
+    if (GetCompositionForEndpointIndex(endpointIndex) == EndpointComposition::kFullFamily)
+    {
+        return DataModel::EndpointInfo(parent, DataModel::EndpointCompositionPattern::kFullFamilyPattern);
+    }
+    if (GetCompositionForEndpointIndex(endpointIndex) == EndpointComposition::kTree)
+    {
+        return DataModel::EndpointInfo(parent, DataModel::EndpointCompositionPattern::kTreePattern);
+    }
+    return std::nullopt;
+}
+
+DataModel::EndpointEntry FirstEndpointEntry(unsigned start_index, uint16_t & found_index)
+{
+    // find the first enabled index after the start index
+    const uint16_t lastEndpointIndex = emberAfEndpointCount();
+    for (uint16_t endpoint_idx = static_cast<uint16_t>(start_index); endpoint_idx < lastEndpointIndex; endpoint_idx++)
+    {
+        if (emberAfEndpointIndexIsEnabled(endpoint_idx))
+        {
+            found_index                            = endpoint_idx;
+            DataModel::EndpointEntry endpointEntry = DataModel::EndpointEntry::kInvalid;
+            endpointEntry.id                       = emberAfEndpointFromIndex(endpoint_idx);
+            auto endpointInfo                      = GetEndpointInfoAtIndex(endpoint_idx);
+            // The endpoint info should have value as this endpoint should be valid at this time
+            VerifyOrDie(endpointInfo.has_value());
+            endpointEntry.info = endpointInfo.value();
+            return endpointEntry;
+        }
+    }
+
+    // No enabled endpoint found. Give up
+    return DataModel::EndpointEntry::kInvalid;
+}
+
+bool SemanticTagsEqual(const DataModel::Provider::SemanticTag & tagA, const DataModel::Provider::SemanticTag & tagB)
+{
+    if (tagA.label.HasValue() != tagA.label.HasValue())
+    {
+        return false;
+    }
+    if (tagA.label.HasValue())
```

Comment:
label is optional and nullable, so probably we need a has_value for both. How about a separate `operator==(const Optional<Nullable<CharSpan>> &a, const Optional<Nullable<CharSpan>> &b)` ?

Spans have data_equal that can be used instead of checking size and strncmp.

Reply:
There is already operator== overload in `Optional` and `Nullable` class, so I added an `operator==(const SemanticTag & a, const SemanticTag & b)` here.